### PR TITLE
Update sidecar to handle blocks outside the 290 block frontier

### LINKED ISSUE #8365: Sidecar bug?
My node's been up since mainnet launch. It ran continuously until yesterday when it crashed for the first time.
I'm running the sidecar as an Ubuntu daemon.
The node is 16 cores 64 GB RAM.

I describe the strange sidecar behaviour below:

Sidecar daemon journalctl logs before crash 1 (all looks good so far):
```
Mar 19 16:59:14 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: INFO:root:Got block data
Mar 19 16:59:15 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: INFO:root:Finished! New tip 897...
Mar 19 17:02:15 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: INFO:root:Fetching block 897...
Mar 19 17:02:17 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: INFO:root:Got block data
Mar 19 17:02:17 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: INFO:root:Finished! New tip 898...
Mar 19 17:05:17 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: INFO:root:Fetching block 898...
```

During crash 1, the error makes sense:

```
Mar 19 17:05:17 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: ERROR:root:<urlopen error [Errno 111] Connection refused>
```
etc...

The crash 1 lasted 18 minutes before the sidecar logs stop throwing errors:

Everything _appeared_ to resume normally...

```
Mar 19 17:23:31 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: ERROR:root:Sleeping for 30s and trying again
Mar 19 17:24:01 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: INFO:root:Fetching block 898...
Mar 19 17:24:12 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: INFO:root:Got block data
Mar 19 17:24:13 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: INFO:root:Finished! New tip 902...
Mar 19 17:27:13 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: INFO:root:Fetching block 902...
```

**BUT**

The logs show the sidecar kept logging `block 902` output (every 3 mins) for *4.5 hours*. Before logging `block 903` for *16 hours*.

^^^ That's the first strange set of behaviour. I don't think the node would have taken hours to re-sync after the crash, because this node is so powerful.

**But the story continues...**

After the aforementioned 16 hours of repeated logging of `block 903`, the sidecar threw errors again. And, indeed, the timing of these errors coincide with a _2nd_ time my node crashed. All understandable so far. 

The problem is, that 2nd crash occurred **7 hours ago**, the node appeared to have resynced quickly after the crash, but the sidecar logs show a consistent stream of errors for the past 7 hours. So although the node restarted itself and resynced correctly, the sidecar has broken somehow and spent 7 hours reporting errors instead of reporting my glorious uptime.

Here are the errors:
First this kind of error (repeated every 3 mins):
```
Mar 20 13:43:40 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: INFO:root:Fetching block 903...
Mar 20 13:43:40 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: ERROR:root:Response seems to be an error! {"errors":[{"message":"Could not obtain transition frontier","path":["block"]}],"data":null}
Mar 20 13:43:40 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: Traceback (most recent call last):
Mar 20 13:43:40 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]:   File "/usr/local/bin/mina-bp-stats-sidecar", line 149, in <module>
Mar 20 13:43:40 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]:     block_data = fetch_block(current_finalized_tip)
Mar 20 13:43:40 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]:   File "/usr/local/bin/mina-bp-stats-sidecar", line 107, in fetch_block
Mar 20 13:43:40 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]:     raise Exception("Response seems to be an error! {}".format(response_body))
Mar 20 13:43:40 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: Exception: Response seems to be an error! {"errors":[{"message":"Could not obtain transition frontier","path":["block"]}],"data":null}
Mar 20 13:43:40 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: ERROR:root:Sleeping for 30s and trying again
```

Then, after some time, it changed to this similar, but slightly different kind of error:

```
Mar 20 13:47:51 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: INFO:root:Fetching block 903...
Mar 20 13:47:55 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: ERROR:root:Response seems to be an error! {"errors":[{"message":"Could not find block in transition frontier with height 903","path":["block"]}],"data":null}
Mar 20 13:47:55 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: Traceback (most recent call last):
Mar 20 13:47:55 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]:   File "/usr/local/bin/mina-bp-stats-sidecar", line 149, in <module>
Mar 20 13:47:55 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]:     block_data = fetch_block(current_finalized_tip)
Mar 20 13:47:55 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]:   File "/usr/local/bin/mina-bp-stats-sidecar", line 107, in fetch_block
Mar 20 13:47:55 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]:     raise Exception("Response seems to be an error! {}".format(response_body))
Mar 20 13:47:55 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: Exception: Response seems to be an error! {"errors":[{"message":"Could not find block in transition frontier with height 903","path":["block"]}],"data":null}
Mar 20 13:47:55 Ubuntu-1804-bionic-64-minimal mina-bp-stats-sidecar[17584]: ERROR:root:Sleeping for 30s and trying again
```

^^^ this one repeated for **6.5 hours**, always trying to fetch 903, even though the node had continued keeping up with the blockchain (now on block 1315 which matches the MinaExplorer).

Upon checking my node and realising the errors, I restarted the sidecar daemon manually and it appears to be working again now.

Unfortunately, I'm worried this has possible resulted in many hours of lost "uptime reporting", where my node was synced but the sidecar has been erroring.

### PR DESCRIPTION
**Pull request by Fitblip**
This PR replaces #8368 and fixes #8365.

TL;DR nodes only keep ~290 blocks in memory at a time, so if we ever get into a situation where we're trying to fetch a block older than that, the sidecar will fail indefinitely until restarted (since it's checking for a block ID that will never successfully be fetched).

This PR fixes this issue by fetching the current block height every time an error is encountered (after ensuring that the node is fully synced). Originally we were also accepting the `CATCHUP` phase before interrogating for blocks, but we should really only be flushing stats on synced nodes so this waits for things to be fully synced before trying to pull down blocks or ship off stats.

### CODE REVIEW

**Review by bkase — APPROVED** [review]
Looks good!

**Comment by bkase** [comment]
!ci-build-me

**Comment by nicolasochem** [comment]
thank you, can we have a new container? https://hub.docker.com/r/minaprotocol/mina-bp-stats-sidecar/tags?page=1&ordering=last_updated
